### ISSUE TITLE
Proposal: Add URL for a rule's documentation to the API

### ISSUE THREAD

**Issue by pmcelhaney** [eslint/eslint#6582]
<h2>The version of ESLint you are using.</h2>

4.12.1

<h2>The problem you want to solve</h2>

AtomLinter has a neat feature that links to the documentation for each rule.

<img width="780" alt="Atom screenshot showing link" src="https://cloud.githubusercontent.com/assets/51504/16542682/8756a1c8-4080-11e6-8a3c-c528b1da91d8.png">

In order for that link to work in plugins, the plugin needs to be [registered in AtomLinter](https://github.com/AtomLinter/linter-eslint/blob/master/src/helpers.js).

<h2>Your take on the correct solution to problem</h2>

Instead of AtomLinter having to maintain a list of known plugins (and other clients having to duplicate the effort) it would be great if the API could allow a rule to specify where its documentation can be found.

It boils down to:
 1. Exposing a rule's meta object via the API (which is a one-liner though several test fixtures are affected).
2. Establishing an optional convention of putting the URL describing a rule in docs.uri / docs.url.
3. Adding a URL to the meta for each of the built-in rules.

<details>
<summary><h3>Original proposal which worked at the plugin level rather than the rule level.</h3></summary>

``` js
module.exports = {
    rules: {
        ...
    },
    ...

    // New property
    getDocURI: function (ruleName) { 
       return `https://myplugin.com/docs/${ruleName}.md`
    }
};

```

The URL would be exposed to the [Node API](http://eslint.org/docs/developer-guide/nodejs-api#linter) via the response from the `verify()` method.

``` js
{
    fatal: false,
    ruleId: "semi",
    severity: 2,
    line: 1,
    column: 23,
    message: "Expected a semicolon.",
    fix: {
        range: [1, 15],
        text: ";"
    },
    docURI:  "http://eslint.org/docs/rules/semi" // <-- New property
}
```

I've [implemented the change](https://github.com/pmcelhaney/eslint/commit/42977491b9a0f1ce235b5fa37f9d57beb1ddd4e7) in ESLint. I'm waiting for the proposal to be accepted so I can open a PR. If it's accepted I'm happy to do the corresponding [update for AtomLinter](https://github.com/AtomLinter/linter-eslint/issues/590) as well. 

</details>

**Comment by platinumazure** [comment]
It'd be good to discuss the details a little. For example, first question I have is why not have the URLs collected in a plugin definition file (same as where plugin rules and configurations are exported)? This also allows the same file to contain a URL for overall documentation, if needed.

**Comment by ilyavolodin** [comment]
This can probably be done very easily through rule's metadata. No code changes necessary. Although we do need an API to retrieve metadata.

**Comment by pmcelhaney** [comment]
Kevin, what you describe is what I have in mind. 

I thought about putting the URL in metadata, but that would become very redundant. Rather define one function at the plugin level.

The concept is more or less proven by AtomLinter. We'd just be moving the knowledge of URIs from AtomLinter to the plugin itself, and expose it through an API that AtomLinter and any other client can consume.

**Comment by pmcelhaney** [comment]
I've updated the issue to clarify exactly what API additions I'm proposing. This would be a new feature without any breaking changes.

**Comment by vitorbal** [comment]
Why would you say putting the URL in the metadata property would be redundant? This seems like the perfect use case for a metadata property.

**Comment by platinumazure** [comment]
I guess I'm not attached one way or the other. I agree it would be great to use some sort of metadata property (either at plugin or rule level) to provide documentation links.

@ilyavolodin Perhaps you want to take a look at the first post of this issue, since it has been updated with a concrete proposal?

**Comment by pmcelhaney** [comment]
I'm in favor of configuring the URI at the plugin level because it saves plugin authors from having to include the URI in every single rule.

**Comment by pmcelhaney** [comment]
First stab at a PR. Feedback much appreciated.

https://github.com/eslint/eslint/pull/6589

**Comment by mysticatea** [comment]
I vote a use of `meta` property.

**Comment by ilyavolodin** [comment]
I understand redundancy argument. However, I feel like "one URL to rule them all" is too inflexible. For example, there might be plugin authors, that don't want to name all of their pages with the same name as rule name. Or they might use some provider that doesn't, for example, support dashes in the urls, but their rules have dashes.
I do understand that you can write complicated implementation of `ruleURI` to support the above listed edge cases, but it feels like recreating the purpose that metadata is supposed to cover.

One other thing to keep in mind, most of the editor addons that might/would use this functionality don't know anything about plugins. So when the error occurs, they would need to parse rule name, extract plugin name out of it, then look up rule URI for it in the API. Vs. just calling API and comparing rule name against the list. No parsing required.

And last thing is, this proposed solution would not work for custom rules, just for plugins. Metadata would work for both.

**Comment by pmcelhaney** [comment]
Thanks, @ilyavolodin. Good thoughts. 

I don't think editors (and their addons) have to deal wirth ESLint plugins. It's all transparent.  A consumer of the Node API shouldn't need to do any parsing. It gets the URI for a reported problem and can do what it wants with it. Anyway, how the URI  is exposed in the Node API is orthaganol to whether the URI is configured by a plugin-level function or rule-level metadata. 

AtomLinter has already proven the concept of using a function to map a ruleId to a URI. All of the functions that have been registered so far have been trivial one-liners. 

Your point about not working for custom rules is valid. However I don't imagine there will be a huge demand for this feature to be used in one-off custom rules. How many people go to the trouble of publishing docs for project-specific rules? And if they do, what's stopping them from moving the rules to  a plugin that can be shared across multiple projects? 

Also there's an advantage to generating a URI on the fly. In the future, the getDocURI() function could take the problem object as an optional  second parameter. That would allow a plugin to generate a URI that points to "smart" documentation, which is tailored to use your actual code snippet in the explanation.

**Comment by pmcelhaney** [comment]
+ @granteagon @relekang @IanVS Looks like you were involved in adding the ESLint plugin docs to the AtomLinter plugin (thank you!). I'm proposing that we allow ESLint plugins to specify where to find their docs and thought you might have something to add to the discussion.

**Comment by granteagon** [comment]
@pmcelhaney, do you mean you are proposing "AtomLinter plugins to specify where to find their docs"?  I did help get the ESLint docs working.

Being out of the loop on the conversation a bit, I think that @ilyavolodin is tending towards a solution I would support.  A blanket "base URL" is probably not a flexible solution.  Meta-data is probably better.

In my opinion, it would be easier to create a build script on the plugins to generate the full URLs for each rule, than it would be to rely on each plugin author to follow a more complex pattern.  Just my two cents.

Great idea.  I've been thinking this should be done for some time now.  I'm happy to help in anyway I can.  Keep me in the loop.  I'm going to loop in @Arcanemagus who helped me implement my code.  He's a pretty smart dude who may be able to help us come to a conclusion.

**Comment by nzakas** [comment]
I agree with @ilyavolodin that this information really belongs in `meta`, probably `meta.docs`. 

On a larger point: all of ESLint is meant so that rules are rules are rules -- we try very hard not to treat rules from the core different from custom rules or different from plugin rules. Any solution we choose must keep that design principle in mind even if it leads to some redundancies.

**Comment by pmcelhaney** [comment]
Okay, thanks. I'll send another PR that puts the config in meta.

**Comment by nzakas** [comment]
@pmcelhaney it would be helpful (and arguably a better use of your time) to propose a design here for review before submitting a new PR.

**Comment by pmcelhaney** [comment]
Thanks @nzakas. Unfortunately by the time I saw your comment I already submitted a new PR. As yourself and others have suggested, I simply exposed the rule metadata through the API (as part of the problem object) and documented `meta.docs.uri` as the standard place to put the URL for a rule's documentation. 

(Aside: I know you've expressed frustration about people reporting bugs / requesting features and not being respectful of the fact that you're all volunteers. Submitting PRs is my way of showing that I'm here to contribute, not just to pester you and expect something for nothing.)

**Comment by nzakas** [comment]
@pmcelhaney and we appreciate the PRs, we're just trying to be respectful of your time by encouraging you to describe what you're thinking of being digging into the code. As this is a change to the core, we have to be more careful about evaluating and accepting it.

**Comment by pmcelhaney** [comment]
From my point of view, the clearest and most expressive way to express my thinking is to just send a PR. Not necessarily a complete PR, but one that hits the key tests, doc updates, and code changes. In this case, the change was relatively small so it was just as easy to do the whole thing (sans adding meta.doc.uri to all of the built-in rules). 

Anyway, if outlining the proposed design in prose works better for you, I'll keep that in mind going forward.

**Comment by Arcanemagus** [comment]
Following along with what @nzakas said earlier, clients (like the `linter-eslint` Atom package) can work with any method, as long as it is consistent between all plugin rules. From what it sounds like the metadata thing will work fine, as long as there is a way to access that information.

**Comment by nzakas** [comment]
@pmcelhaney we are just trying to be respectful of everyone's time. It's generally a lot less time to write up a proposal than to implement, and likewise, it's a lot less time for the team to be able to review a narrative proposal of what you're thinking than to look through the code and try to figure it out.

**Comment by pmcelhaney** [comment]
Still interested in finishing if this proposal is accepted. 

It boils down to:

1. Exposing a rule's meta object via the API (which is a one-liner though several test fixtures are affected). 
2. Establishing an optional  convention of putting the URL describing a rule in docs.uri / docs.url.
3. Adding a URL  to the meta for each of the built-in rules.

**Comment by granteagon** [comment]
@pmcelhaney are you still working on it?

**Comment by pmcelhaney** [comment]
The work is done, just awaiting feedback on the design.

**Comment by granteagon** [comment]
@pmcelhaney I don't see a PR.  Where can I check it out?

**Comment by pmcelhaney** [comment]
https://github.com/pmcelhaney/eslint/commit/42977491b9a0f1ce235b5fa37f9d57beb1ddd4e7

**Comment by IanVS** [comment]
[Adding meta to the problem object](https://github.com/pmcelhaney/eslint/commit/42977491b9a0f1ce235b5fa37f9d57beb1ddd4e7#diff-b6559e6b095a292d244b6b7bae986291R986) would be great for tools like Atom's linter-eslint.  I'm 👍 for this change.

**Comment by jfmengels** [comment]
Same as @IanVS.

Having the rule's meta would allow quite a few nice things I think.
For instance, having the category of the rule easily available to the editor would allow it to for instance change the color of the error based on whether it's a "possible error" or a "stylistic issue" (should this field exist). This is a bad example as there is already two colors for warn and error, but somehow the appearance could differ.

**Comment by ilyavolodin** [comment]
I think adding URL to the meta is the right way to go. What I'm not 100% sure about is if we want to include metadata in the `context.report`, or if we should provide another API that would return list of all rules with metadata, that can be used for a look-up. I feel like in a large number of files run including metadata with every error might result in performance problems or in some extreme cases, in out of memory errors.

**Comment by IanVS** [comment]
> out of memory errors

How many linting errors are you getting?!?!  😝 

I panic-deleted my original comment that @jfmengels was echoing, wherein I said it would be great to have the meta (at least the docs meta) available on the problem object for tools like Atom's linter-eslint.  I thought maybe it would only be available to consumers of CLIEngine, which linter-eslint doesn't use, but it seems like it would be in the report as well, so it's all good.

**Comment by IanVS** [comment]
>or if we should provide another API that would return list of all rules with metadata

That would be sub-optimal for linter-eslint, as we use the cli itself, and not CLIEngine.

**Comment by ilyavolodin** [comment]
@IanVS We've had reports of people running out of memory generating json reports. If you run ESLint with strict config on a legacy codebase, it's not surprising to see hundred of thousands of errors. But maybe it's just preemptive optimization and we shouldn't worry about it until we see some evidence of issues. In general I'm fine with @pmcelhaney implementation if everyone else is OK with it.

> That would be sub-optimal for linter-eslint, as we use the cli itself, and not CLIEngine.

Don't really want to get into this discussion, but that seems like a very strange decision.

**Comment by pmcelhaney** [comment]
Good point about memory. Shouldn't be a problem if we [pass the meta object by reference](
https://github.com/pmcelhaney/eslint/commit/42977491b9a0f1ce235b5fa37f9d57beb1ddd4e7#diff-b6559e6b095a292d244b6b7bae986291R459).

**Comment by kaicataldo** [comment]
Where does this issue stand? Looks like a significant amount of work was done, but there hasn't been any activity in quite a long time.

**Comment by ilyavolodin** [comment]
Meta object has been added to all core rules. We also now have API to retrieve metadata (and rules). I still think it would be nice to add metadata to the report object. So that would be the last thing that would need to be added. I can champion this proposal.

**Comment by kaicataldo** [comment]
@eslint/eslint-team It looks like we have a champion and 2 :+1:. Anyone else interested in supporting this?

**Comment by not-an-aardvark** [comment]
I'm not sure I understand the proposal. How would the metadata be added to the report object?

**Comment by not-an-aardvark** [comment]
Also, can't integrations access the metadata anyway?

```js
const ruleId = message.ruleId;
const rule = linter.getRules()[ruleId];
const url = rule.meta.url;
```

**Comment by IanVS** [comment]
To verify, `linter.getRules()` returns plugin rules as well as core rules, correct?

**Comment by not-an-aardvark** [comment]
I think so, provided that the plugins have been loaded on the given `linter` object.

**Comment by pmcelhaney** [comment]
If there's a way to get the meta object we probably don't need to change the source code, but points 2 and 3 remain. 

Accessing the metadata via existing APIs
Establishing an optional convention of putting the URL describing a rule in docs.uri / docs.url.
Adding a URL to the meta for each of the built-in rules.

**Comment by platinumazure** [comment]
Okay, I commented on this long ago and then never followed up- sorry.

I'm :+1: to using a property in rule metadata for this. If someone wanted to expose URLs in the plugin definition file, they could do so by requiring the rule and getting the URI from its metadata. So I withdraw any objection I might have raised earlier.

I think TSC needs to review a proposal before this can be accepted. I could write up a summary in a day or two, but if someone beats me to it, all the better.

**Comment by pmcelhaney** [comment]
I’ll write up a formal proposal if someone sends me an example or two of what a good TSC proposal looks like.

**Comment by ilyavolodin** [comment]
@pmcelhaney You can use our [Change Request](https://eslint.org/docs/developer-guide/contributing/changes) template and add a Summary and a Question for TSC at the end as well, so it would be easier for TSC members to understand at a glances what the problem is and what they are suppose to vote on.

**Comment by pmcelhaney** [comment]
I've updated the description to match the change request template. I don't have any questions for TSC.  The title is the summary; the entire proposal is less than 100 words.

**Comment by Arcanemagus** [comment]
Just a little note, the code you used to be linking to in `linter-eslint` now lives in [`eslint-rule-documentation`](https://github.com/jfmengels/eslint-rule-documentation/). Thanks for getting this filed and the process started @pmcelhaney!

**Comment by ilyavolodin** [comment]
**TSC Summary**: This proposal suggests adding docs.uri/docs.url to meta property of each built-in rule with the link to rule's documentation.

**TSC Question**: Do we accept this proposal?

**Comment by not-an-aardvark** [comment]
This proposal was accepted in [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/3ef527b67712b5cecd0e2231dfb1a8a59a052155/notes/2017/2017-12-07.md#proposal-add-url-for-a-rules-documentation-to-the-api). Additionally, we decided to add a `CLIEngine#getRules` method to allow integrations that use `CLIEngine` to access the rules (as opposed to integrations that only use `Linter`).

**Comment by Arcanemagus** [comment]
> Additionally, we decided to add a `CLIEngine#getRules` method to allow integrations that use `CLIEngine` to access the rules (as opposed to integrations that only use `Linter`).

Yay! No more [undocumented hack](https://github.com/AtomLinter/linter-eslint/blob/7f791653480ce30c1ffe9ab99c6c84e34fe7e7fe/src/worker.js#L67-L69) needed!

**Comment by not-an-aardvark** [comment]
I think this issue can be closed now that we've added `CLIEngine#getRules` and established the `meta.docs.url` convention for rules.
